### ISSUE TITLE
Cannot set PREFIX to a cons cell (PREFIX . DESCRIPTION) when using map! with nested prefixes

### ISSUE THREAD

**Issue by hugonobrega** [hlissner/doom-emacs#4074]
I am trying to create keybindings using the "map!" macro with nested :prefix clauses. 
In order for which-key to be more helpful, I would like to use :prefix with the cons cell (prefix . description), as described in the help for map!

This works fine for a single :prefix, or for nested :prefixes where only the top-most level uses the cons cell format, but fails with error "Wrong type argument: stringp" if I try to use the cons cell format in the inner :prefix clauses.

It works if I replace :prefix by :prefix-map, but the documentation is pretty adamant that I should not do that in my own config file.

**Steps to reproduce:**

Evaluate
```
(defun test ()
  (interactive)
  (message "Hi!"))
```
Then evaluate each of the following blocks in turn (this assumes you have a Hyper key bound, otherwise please change H-a to some free key in your setup):
```
;; Works:
(map! :prefix "H-a"
      "b" #'test
      (:prefix "H-a"
       "b" #'test))
```

```
;; Also works:
(map! :prefix ("H-a" . "First Description")
      "b" #'test
      (:prefix "H-a"
       "b" #'test))
```

```
;; Also works, but seems to be advised against:
(map! :prefix ("H-a" . "First Description")
      "b" #'test
      (:prefix-map ("H-a" . "Second Description")
       "b" #'test))
```

```
;; Fails (backtrace below):
(map! :prefix ("H-a" . "First Description")
      "b" #'test
      (:prefix ("H-a" . "Second Description")
       "b" #'test))
```

**System information:**
<details><pre>
SYSTEM  type       gnu/linux
        config     x86_64-pc-linux-gnu
        shell      /bin/bash
        uname      Linux 5.8.11-1-MANJARO #1 SMP PREEMPT Wed Sep 23 14:35:40 UTC 2020 x86_64
        path       (~/programs/doom-emacs/bin/ ~/.local/bin /usr/local/bin /usr/bin /bin /usr/local/sbin /usr/bin/site_perl /usr/bin/vendor_perl /usr/bin/core_perl /var/lib/snapd/snap/bin ~/programs/texlive/2020/bin/x86_64-linux /usr/libexec/emacs/28.0.50/x86_64-pc-linux-gnu/)
EMACS   dir        ~/programs/doom-emacs/
        version    28.0.50
        build      out 04, 2020
        buildopts  --prefix=/usr --with-nativecomp --with-xwidgets
        features   XPM JPEG TIFF GIF PNG RSVG CAIRO SOUND GPM DBUS GSETTINGS GLIB NOTIFY INOTIFY ACL GNUTLS LIBXML2 FREETYPE HARFBUZZ ZLIB TOOLKIT_SCROLL_BARS GTK3 X11 XDBE XIM MODULES NATIVE_COMP THREADS XWIDGETS LIBSYSTEMD JSON PDUMPER LCMS2
        traits     (daemon server-running chemacs envvar-file)
DOOM    dir        ~/.doom.d/
        version    2.0.9
        build      HEAD -> develop 7833cee77 2020-10-09 18:34:14 -0400
        elc-files  0
        modules    (:completion company ivy :ui doom doom-dashboard hl-todo hydra modeline ophints (popup +defaults) vi-tilde-fringe :editor (evil +everywhere) file-templates fold snippets :emacs dired electric ibuffer :tools (eval +overlay) lookup magit pdf :lang emacs-lisp (haskell +dante) latex markdown (org +hugo +ipython +roam +noter) python sh :email (mu4e +gmail) :config literate (default +bindings +smartparens))
        packages   ((ace-window) (eyebrowse) (undo-tree) (transpose-frame) (mu4e-conversation) (try) (swiper) (org-ref) (ivy-bibtex) (pipenv :disable t) (nose :disable t) (python-pytest :disable t) (visual-regexp) (visual-regexp-steroids))
        unpin      (n/a)
        elpa       (n/a)
</pre></details>

**Backtrace:**
<details><pre>
Debugger entered--Lisp error: (wrong-type-argument stringp [16777313 16777313])
  general-extended-def-:which-key(nil global [16777313 16777313] (:def nil :ignore t :which-key "Second Description") (:wk-match-keys t :wk-match-binding t :wk-full-keys t :keymaps (global) :states (nil) :prefix [16777313 16777313]))
  #f(compiled-function (keywords &optional alter-def) #<bytecode -0x5564feae90ed100>)((:which-key :wk :properties :repeat :jump))
  general--run-extended-def-functions(nil global [16777313 16777313] (:def nil :ignore t :which-key "Second Description") (:wk-match-keys t :wk-match-binding t :wk-full-keys t :keymaps (global) :states (nil) :prefix [16777313 16777313]))
  general--parse-def(nil global [16777313 16777313] (:ignore t :which-key "Second Description") (:wk-match-keys t :wk-match-binding t :wk-full-keys t :keymaps (global) :states (nil) :prefix [16777313 16777313]))
  general--parse-maps(nil global ([16777313 16777313] (:ignore t :which-key "Second Description") [16777313 16777313 98] test) (:wk-match-keys t :wk-match-binding t :wk-full-keys t :keymaps (global) :states (nil) :prefix [16777313 16777313]))
  general--define-key(nil global ([16777313 16777313] (:ignore t :which-key "Second Description") [16777313 16777313 98] test) nil nil (:wk-match-keys t :wk-match-binding t :wk-full-keys t :keymaps (global) :states (nil) :prefix [16777313 16777313]))
  general--delay((or (boundp 'global) (and (memq 'global '(local global)) (if 'nil (and (boundp 'evil-state-properties) (assq 'nil evil-state-properties)) t))) (general--define-key 'nil 'global '([16777313 16777313] (:ignore t :which-key "Second Description") [16777313 16777313 98] test) 'nil 'nil '(:wk-match-keys t :wk-match-binding t :wk-full-keys t :keymaps (global) :states (nil) :prefix [16777313 16777313])) after-load-functions t nil "general-define-key-in-global296")
  general-define-key(:prefix [16777313 16777313] "" (:ignore t :which-key "Second Description") "b" test)
  (progn (general-define-key :prefix "H-a" "" (list :ignore t :which-key "First Description") "b" #'test) (general-define-key :prefix (general--concat nil "H-a" "H-a") "" (list :ignore t :which-key "Second Description") "b" #'test))
  elisp--eval-last-sexp(nil)
  #<subr eval-last-sexp>(nil)
  apply(#<subr eval-last-sexp> nil)
  eval-last-sexp(nil)
  eros-eval-last-sexp(nil)
  funcall-interactively(eros-eval-last-sexp nil)
  command-execute(eros-eval-last-sexp)
</pre></details>

**Comment by hugonobrega** [comment]
A workaround:
```
;; Works:
(map! :prefix ("H-a" :which-key "First Description")
      "b" #'test
      (:prefix ("H-a" :which-key "Second Description")
       "b" #'test))
```

**Comment by hlissner** [comment]
This was due to a bug in `map!` and was fixed in d9683d1. Let me know if that isn't the case and I'll reopen it. Thanks for bringing it to my attention!

**Comment by hugonobrega** [comment]
Works perfectly now, many thanks for this (and for Doom in general)!
